libs/framework/src/service_registry.c

```diff
@@ -277,13 +277,12 @@ celix_status_t serviceRegistry_unregisterService(service_registry_pt registry, b
         service_reference_pt ref = refsMap != NULL ?
                                    hashMap_get(refsMap, (void*)registration->serviceId) : NULL;
         if (ref != NULL) {
-            serviceReference_invalidate(ref);
+            serviceReference_invalidateCache(ref);
         }
     }
     hashMapIterator_destroy(iter);
+    serviceRegistration_invalidate(registration);
 	celixThreadRwlock_unlock(&registry->lock);
```

Comment:
Between unlock and invalidate, there is a window of race condition: `serviceRegistry_getServiceReference` will return a reference containing a still valid registration (but to be invalidated soon), through which we can get a service instance, which has no chance of ungetting once the registration is invalidated.

Reply:
Just to be sure. There **was** a window for a race condition, right? 

Because now that the serviceRegistration_invalidate has moved to before the unlock, this should not happen anymore.

Reply:
Yes. If serviceRegistry_getServiceReference happens before this critical section, the valid registration will be invalidated by the critical section; if it happens after this critical section, then the returned reference can only bind to invalid registration. The lock guarantees that these are the only possible cases.